clinica/iotools/converters/ukb_to_bids/ukb_utils.py

```diff
@@ -0,0 +1,369 @@
+
+    source_file = fs.open(fs.ls(sourcedata_dir)[0], mode="rb")
+    target_file = fs.open(bids_filename, mode="wb")
+
+    with source_file as sf, target_file as tf:
+        tf.write(sf.read())
+
+    source_basename = Path(Path(Path(fs.ls(sourcedata_dir)[0]).stem).stem)
+    target_basename = Path(bids_filename.stem).stem
+
+    # The following part adds the sidecar files related to the nifti with the same name: it can be tsv or json files.
+    # It may or may not be used, since there might not be any sidecars.
+    sidecar_dir = sourcedata_dir.parent
+    for source_sidecar in sidecar_dir.rglob(f"{source_basename}*"):
+        target_sidecar = Path.joinpath(bids_filename.parent, target_basename).with_name(
+            f"{target_basename}{source_sidecar.suffix}"
+        )
+        source_file = fs.open(source_sidecar, mode="rb")
+        target_file = fs.open(target_sidecar, mode="wb")
+        with source_file as sf, target_file as tf:
+            tf.write(sf.read())
+
+
+def write_bids(
+    to: PathLike,
+    participants: DataFrame,
+    sessions: DataFrame,
+    scans: DataFrame,
+    dataset_directory: PathLike,
+) -> List[PathLike]:
+    from pathlib import Path
+
+    from fsspec.implementations.local import LocalFileSystem
+
+    from clinica.iotools.bids_dataset_description import BIDSDatasetDescription
+
+    to = Path(to)
+    fs = LocalFileSystem(auto_mkdir=True)
+
+    # Ensure BIDS hierarchy is written first.
+    with fs.transaction:
+        with fs.open(to / "dataset_description.json", "w") as dataset_description_file:
+            BIDSDatasetDescription(name="UKB").write(to=dataset_description_file)
+
+        with fs.open(to / "participants.tsv", "w") as participant_file:
+            write_to_tsv(participants, participant_file)
+
+    for participant_id, data_frame in sessions.groupby(["participant_id"]):
+        session = data_frame.droplevel("participant_id")
+        session_filepath = to / participant_id / f"{participant_id}_sessions.tsv"
+        with fs.open(session_filepath, "w") as sessions_file:
+            write_to_tsv(session, sessions_file)
+
+    scans = scans.set_index(["bids_full_path"], verify_integrity=True)
+    for bids_full_path, metadata in scans.iterrows():
+        install_nifti(
+            zipfile=str(dataset_directory) + "/" + metadata["source_zipfile"],
+            filename=metadata["source_filename"],
+            bids_path=to / bids_full_path,
+        )
+        install_json(
+            zipfile=str(dataset_directory) + "/" + metadata["source_zipfile"],
+            filename=metadata["json"],
+            bids_path=to / metadata["bids_sidecar_path"],
+        )
+        install_sidecars(
+            zipfile=str(dataset_directory) + "/" + metadata["source_zipfile"],
+            filename=metadata["sidecars"],
+            bids_path=to / metadata["bids_sidecar_path"],
+        )
+    return
+
+
+def install_nifti(zipfile: str, filename: str, bids_path: str) -> None:
```

Comment:
Might be just me, but I find the names (esp. the `install` prefix) a bit misleading.
It looks like these functions are extracting files from an archive and writing them to some location.
Also, for DRYness, it looks like you could easily combine `install_nifti` with `install_json` since they have almost exactly the same code.

Reply:
Yes, i was thinking about it as well. I may not combine `install_nifti`and `install_json`, but i will probably combine `install_json`and `install_sidecars`

Reply:
I'd agree with @NicolasGensollen on this one,  as the content is almost identical, maybe rename the function to something more generic as `extract_data` and reuse.

Reply:
Done !